src/lib.rs

```diff
@@ -253,9 +253,13 @@ impl<'t> MachineDescriptor<'t> {
 
         Ok(Machine {
             nickname: self.nickname,
+            // if not defined, set public dns to be the public ip
+            public_dns: self
+                .public_dns
+                .map(String::from)
+                .unwrap_or(self.public_ip.to_string()),
```

Comment:
This should technically be an `unwrap_or_else(||`, otherwise you'll end up allocating an extra string on each invocation.

Reply:
I've tried that but I'm getting the following error:
```
error[E0382]: borrow of moved value: `self`
   --> src/lib.rs:260:33
    |
257 |               public_dns: self
    |  _________________________-
258 | |                 .public_dns
    | |___________________________- value moved here
259 |                   .map(String::from)
260 |                   .unwrap_or_else(|| self.public_ip.to_string()),
    |                                   ^^ ---- borrow occurs due to use in closure
    |                                   |
    |                                   value borrowed here after move
    |
    = note: move occurs because `self.public_dns` has type `std::option::Option<std::string::String>`, which does not implement the `Copy` trait
```

I'm not sure how to fix this.

Reply:
Ah, yes, that's because you're moving out of `self` in the same constructor (when you construct all the other fields). If you just do it separately above, it should be fine:
```rust
let public_dns = self
                .public_dns
                .unwrap_or(self.public_ip.to_string());
```
and then use `public_dns` instead of `self.public_dns` below.

Reply:
I had to first move `self.public_ip` to its own variable, and only then compute `self.public_dns`:

```rust
        let public_ip = self.public_ip;
        // if not defined, set public dns to be the public ip
        let public_dns = self
            .public_dns
            .map(String::from)
            .unwrap_or_else(|| public_ip.clone());
```

I guess there was no need to also `public_dns` outside of `Machine`'s constructor.

Reply:
Ah, yes, that's because Rust only sees the closure borrow _all of `self`_, not just `self.public_ip`. This is because the compiler's understanding of split/partial borrows is limited to a single scope (it can't, for example, see "through" a closure for what fields it is accessing).

Reply:
That makes sense, thanks! I've also had this problem with methods that borrow self: even though a single field is accessed, all self is borrowed. I guess it's the same problem. Do you know if there are any plans to address this?

Reply:
I think the thing to watch is probably https://github.com/rust-lang/rfcs/issues/1215, though not sure if that'll solve this particular pain point :shrug:

Reply:
Thanks!